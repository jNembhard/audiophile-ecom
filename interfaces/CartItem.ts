interface CartItem {
  id: number;
  shortName: string;
  cartImg: string;
  price: number;
  quantity: number;
}

export default CartItem;
