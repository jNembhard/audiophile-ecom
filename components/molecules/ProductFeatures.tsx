import {
  Stack,
  Box,
  Heading,
  Text,
  List,
  ListItem,
  SimpleGrid,
} from "@chakra-ui/react";
import { IncludedItem } from "@/interfaces/Product";

type Props = {
  features: string | undefined;
  includedItems: IncludedItem[] | undefined;
};

const ProductFeatures = (props: Props) => {
  return (
    <Stack
      direction={{ base: "column", lg: "row" }}
      spacing={{ base: "5.5rem", lg: "7.5rem" }}
      mt={{ base: "5.5rem", sm: "7.5rem", lg: "10rem" }}
    >
      <Box flexBasis={{ lg: "55%" }}>
        <Heading id="features" fontSize={{ base: "1.5rem", md: "2rem" }}>
          Features
        </Heading>
        <Text
          aria-labelledby="features"
          mt={{ base: "1.5rem" }}
          whiteSpace="pre-line"
        >
          {props.features}
        </Text>
      </Box>
      <Stack
        direction={{ base: "column", sm: "row", lg: "column" }}
        spacing={{ base: "1.5rem", sm: "6rem", md: "12rem", lg: "2rem" }}
      >
        <Heading fontSize={{ base: "1.5rem", md: "2rem" }}>In the Box</Heading>
        <List mt={{ base: "1.5rem" }}>
          {props.includedItems?.map((includedItem) => (
            <ListItem key={includedItem.item} fontSize="0.9375rem" mb="0.5rem">
              <SimpleGrid gridTemplateColumns="2rem 1fr">
                <Box as="span" color="rawSienna" fontWeight="bold">
                  {includedItem.quantity}x
                </Box>
                <Box
                  as="span"
                  color="rgba(0,0,0,0.5)"
                  textTransform="capitalize"
                  textAlign="left"
                >
                  {includedItem.item}
                </Box>
              </SimpleGrid>
            </ListItem>
          ))}
        </List>
      </Stack>
    </Stack>
  );
};

export default ProductFeatures;
