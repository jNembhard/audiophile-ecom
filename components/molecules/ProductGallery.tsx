import { Box, Image, SimpleGrid } from "@chakra-ui/react";
import { CategoryImage } from "@/interfaces/Product";

type Props = {
  first?: CategoryImage | undefined;
  second?: CategoryImage | undefined;
  third?: CategoryImage | undefined;
  altText?: CategoryImage | undefined;
};

const ProductGallery = (props: Props) => {
  return (
    <SimpleGrid
      gridTemplateColumns={{ base: "1fr", sm: "40% 1fr" }}
      templateAreas={{ sm: '"a c" "b c"' }}
      mt={{ base: "5.5rem", sm: "7.5rem", lg: "10rem" }}
      gap={{ base: "1.25rem" }}
    >
      <Box as="picture" gridArea={{ sm: "a" }}>
        <source media="(min-width: 62em)" srcSet={props.first?.desktop} />
        <source media="(min-width: 30em)" srcSet={props.first?.tablet} />
        <Image
          src={props.first?.mobile}
          objectFit="cover"
          width="100%"
          height="100%"
          borderRadius="0.5rem"
          alt={props.first?.altText}
        />
      </Box>
      <Box as="picture" gridArea={{ sm: "b" }}>
        <source media="(min-width: 62em)" srcSet={props.second?.desktop} />
        <source media="(min-width: 30em)" srcSet={props.second?.tablet} />
        <Image
          src={props.second?.mobile}
          objectFit="cover"
          width="100%"
          height="100%"
          borderRadius="0.5rem"
          alt={props.second?.altText}
        />
      </Box>
      <Box as="picture" gridArea={{ sm: "c" }}>
        <source media="(min-width: 62em)" srcSet={props.third?.desktop} />
        <source media="(min-width: 30em)" srcSet={props.third?.tablet} />
        <Image
          src={props.third?.mobile}
          objectFit="cover"
          width="100%"
          height="100%"
          borderRadius="0.5rem"
          alt={props.third?.altText}
        />
      </Box>
    </SimpleGrid>
  );
};

export default ProductGallery;
