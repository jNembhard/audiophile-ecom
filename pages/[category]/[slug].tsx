import Head from "next/head";
import Params from "../../interfaces/Params";
import { Product } from "../../interfaces/Product";
import { getProductPaths, getProductBySlug } from "../../utils/products";
import { GetStaticPaths, GetStaticProps } from "next";
import ProductPageTemplate from "../../components/templates/ProductPageTemplate";

type Props = { product: Product };

const ProductPage: React.FC<Props> = (props: Props) => {
  return (
    <div>
      <Head>
        <title>{`${props.product.name} | Audiophile`}</title>
        <meta name="description" content="Generated by create next app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <ProductPageTemplate product={props.product} />
    </div>
  );
};

export const getStaticPaths: GetStaticPaths = async () => {
  const paths = getProductPaths().map((path) => ({
    params: { category: path.category, slug: path.slug },
  }));

  return { paths, fallback: false };
};

export const getStaticProps: GetStaticProps = async (context) => {
  const params = context.params as Params;
  const product: Product | undefined = getProductBySlug(params.slug);

  return { props: { product } };
};

export default ProductPage;
