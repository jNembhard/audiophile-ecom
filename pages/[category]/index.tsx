import Head from "next/head";
import { useRouter } from "next/router";
import { GetStaticPaths, GetStaticProps } from "next";
import { getProductsByCategory, getCategories } from "../../utils/products";
import { Product } from "../../interfaces/Product";
import CategoryTemplate from "../../components/templates/CategoryTemplate";
import Params from "../../interfaces/Params";

type Props = { products: Product[] };

const CategoryPage = (props: Props) => {
  const {
    query: { category },
  } = useRouter();
  return (
    <div>
      <Head>
        <title>{`Audiophile Product Category`}</title>
        <meta name="description" content="Generated by create next app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <CategoryTemplate products={props.products} />
    </div>
  );
};

export default CategoryPage;

export const getStaticPaths: GetStaticPaths = async () => {
  const categories: string[] = getCategories();
  const paths = categories.map((category) => ({ params: { category } }));

  return { paths, fallback: false };
};

export const getStaticProps: GetStaticProps = async (context) => {
  const params = context.params as Params;
  const products: Product[] = getProductsByCategory(params.category);

  return { props: { products } };
};
