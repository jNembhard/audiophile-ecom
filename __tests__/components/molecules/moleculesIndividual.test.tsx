import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { mockAllIsIntersecting } from "react-intersection-observer/test-utils";
import { ChakraWrapper } from "@/test_utils/ChakraWrapper";
import { mockNextUseRouter } from "@/test_utils/mock_router/mockrouter";
import { getStaticProps } from "@/pages/[category]/[slug]";
import { GetStaticPropsContext } from "next";
import ProductGallery from "@/components/molecules/ProductGallery";
import ProductFeatures from "@/components/molecules/ProductFeatures";
import { IncludedItem } from "@/interfaces/Product";

describe("ProductGallery", () => {
  mockNextUseRouter({
    route: "/xx99-mark-one-headphones",
    pathname: "/xx99-mark-one-headphones",
    query: "xx99-mark-one-headphones",
    asPath: "",
  });

  it("should render the ProductGallery Component", async () => {
    const context = {
      params: {
        category: "headphones",
        slug: "xx99-mark-one-headphones",
      },
    };
    const value: any = await getStaticProps(context as GetStaticPropsContext);
    const productGallery = value.props.product.gallery;
    console.log(productGallery);

    render(
      <ChakraWrapper>
        <ProductGallery {...productGallery} />
      </ChakraWrapper>
    );

    mockAllIsIntersecting(true);

    const xx59Headphones = [
      "XX99 Mark I Headphones-1",
      "XX99 Mark I Headphones-2",
      "XX99 Mark I Headphones-3",
    ];

    xx59Headphones.forEach((xx59Headphone) => {
      const headphones = screen.getByAltText(xx59Headphone);
      expect(headphones).toBeInTheDocument();
    });
  });
});

describe("ProductFeatures", () => {
  mockNextUseRouter({
    route: "/xx99-mark-one-headphones",
    pathname: "/xx99-mark-one-headphones",
    query: "xx99-mark-one-headphones",
    asPath: "",
  });

  it("should render the ProductFeature Component with dynamic data", async () => {
    const context = {
      params: {
        category: "headphones",
        slug: "xx99-mark-one-headphones",
      },
    };
    const value: any = await getStaticProps(context as GetStaticPropsContext);
    const features = value.props.product.features;
    const includedItems = value.props.product.includedItems;

    render(
      <ChakraWrapper>
        <ProductFeatures features={features} includedItems={includedItems} />
      </ChakraWrapper>
    );

    mockAllIsIntersecting(true);

    const xx99Text =
      "As the headphones all others are measured against, the XX99 Mark I demonstrates over five decades of audio expertise, redefining the critical listening experience. This pair of closed-back headphones are made of industrial, aerospace-grade materials to emphasize durability at a relatively light weight of 11 oz. From the handcrafted microfiber ear cushions to the robust metal headband with inner damping element, the components work together to deliver comfort and uncompromising sound. Its closed-back design delivers up to 27 dB of passive noise cancellation, reducing resonance by reflecting sound to a dedicated absorber. For connectivity, a specially tuned cable is includes with a balanced gold connector.";

    const xx99Items: IncludedItem[] = [
      { quantity: 1, item: "Headphone unit" },
      { quantity: 2, item: "Replacement earcups" },
      { quantity: 1, item: "User manual" },
      { quantity: 1, item: "3.5mm 5m audio cable" },
    ];

    const featureText = screen.getByLabelText(/Features/);
    expect(featureText).toHaveTextContent(xx99Text);

    xx99Items.forEach((xx99Item: IncludedItem) => {
      let item = screen.getByText(xx99Item.item);
      let quantity = screen.getAllByText(
        `${xx99Item.quantity}x`
      )[0] as HTMLSpanElement;

      expect(item).toBeInTheDocument();
      expect(quantity).toMatchSnapshot(`<span class="css-1vroocg">1x</span>`);
    });
  });
});
