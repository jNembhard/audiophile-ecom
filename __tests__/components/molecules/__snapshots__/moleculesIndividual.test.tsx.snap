// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductFeatures should render the ProductFeature Component with dynamic data: <span class="css-1vroocg">1x</span> 1`] = `
<span
  class="css-1vroocg"
>
  1
  x
</span>
`;

exports[`ProductFeatures should render the ProductFeature Component with dynamic data: <span class="css-1vroocg">1x</span> 2`] = `
<span
  class="css-1vroocg"
>
  2
  x
</span>
`;

exports[`ProductFeatures should render the ProductFeature Component with dynamic data: <span class="css-1vroocg">1x</span> 3`] = `
<span
  class="css-1vroocg"
>
  1
  x
</span>
`;

exports[`ProductFeatures should render the ProductFeature Component with dynamic data: <span class="css-1vroocg">1x</span> 4`] = `
<span
  class="css-1vroocg"
>
  1
  x
</span>
`;
